infra/Pos/Communication/Relay/Types.hs

```diff
@@ -44,10 +45,17 @@ instance Buildable PropagationMsg where
     build (DataOnlyPM conts) =
         Buildable.build conts
 
--- | Queue of InvMsges which should be propagated.
-type RelayPropagationQueue = TBQueue (Microsecond, PropagationMsg)
+data RelayContext m = RelayContext
```

Comment:
> So RelayContext contains everything you might need for implementation. If at some point need to use something else, you can add it to RelayContext.

But this is unacceptable, since the implementation comes from time-warp. The implementation *must* mention the application monad `m`.

The goal is that the data necessary to do relaying is determined by time-warp, and kept abstract. cardano-sl types morally should not make any claims about what the underlying relay mechanism requires (how could it possibly know?); it merely instantiates it (allocates the resource, if you will) and then uses it to discharge some reader transformer. The only claims it makes about the implementation is that the effect carrier `m` supports everything that the relayer needs. GHC will take care of that when it solves constraints.